### ISSUE TITLE
pyff choking on XADES-signed EntityDescriptor

### ISSUE THREAD

**Issue by rhoerbe** [leifj/pyFF#86]
Due to a misconfiguration of my setup a stylesheet to remove signatures in pyff's input directory was not applied and pyff had to eat documents with a (valid) XADES signature, causing an exception:

```
DEBUG:root:transform: http://www.w3.org/2000/09/xmldsig#enveloped-signature
DEBUG:root:using hash algorithm sha256
DEBUG:root:<sha256 HASH object @ 0x19b6f60>
DEBUG:root:using digest JvmW5vKjaTEVHzOdiC/H3HSGNocGamY9sDeU86ld6TA= (sha256) for ref 
DEBUG:root:<ds:DigestValue xmlns:ds="http://www.w3.org/2000/09/xmldsig#" xmlns:alg="urn:oasis:names:tc:SAML:metadata:algsupport" xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" xmlns:mdattr="urn:oasis:names:tc:SAML:metadata:attribute" xmlns:mdrpi="urn:oasis:names:tc:SAML:metadata:rpi" xmlns:mdui="urn:oasis:names:tc:SAML:metadata:ui" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:xrd="http://docs.oasis-open.org/ns/xri/xrd-1.0">JvmW5vKjaTEVHzOdiC/H3HSGNocGamY9sDeU86ld6TA=</ds:DigestValue>
DEBUG:root:Looking for #SignedProperties-f0afeff-1 using id attribute 'ID'
DEBUG:root:Looking for #SignedProperties-f0afeff-1 using id attribute 'id'
----------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/pyff/__init__.py", line 53, in main plumbing(p).process(md, state={'batch': True, 'stats': {}})
  File "/usr/lib/python2.6/site-packages/pyff/pipes/__init__.py", line 165, in process 
    self._process(req)
  File "/usr/lib/python2.6/site-packages/pyff/pipes/__init__.py", line 184, in _process
    ot = pipe(req, *opts)
  File "/usr/lib/python2.6/site-packages/pyff/pipes/builtins.py", line 581, in sign
    xmlsec.sign(req.t, key_file, cert_file, **opts)
  File "/usr/lib/python2.6/site-packages/xmlsec/__init__.py", line 575, in sign
    _process_references(t, sig, return_verified=False, sig_path=sig_path)
  File "/usr/lib/python2.6/site-packages/xmlsec/__init__.py", line 294, in _process_references
    obj = _remove_child_comments(_get_by_id(ct, uri[1:]))
  File "/usr/lib/python2.6/site-packages/xmlsec/__init__.py", line 273, in _remove_child_comments
    for c in t.iter():
AttributeError: 'NoneType' object has no attribute 'iter'
----------------------------------------------------------------
```

While the workaround is to remove the signature, it would be nice if pyff would handle this gracefully.

**Comment by leifj** [comment]
can you give me a minimal example that reproduces this? I note that the code breaks in sign

**Comment by rhoerbe** [comment]
https://github.com/rhoerbe/PVZDpolman/blob/master/PolicyManager/tests/testdata/PAT/02/idpExampleCom.xml

**Comment by leifj** [comment]
There are actually problems with that xml - Id isn't globally unique, but I don't think that is the issue. Can you share your pipeline too?

**Comment by leifj** [comment]
can you share your pipeline?

**Comment by rhoerbe** [comment]
```
- load:
  - /var/lib/pyff/input
- select
- xslt:
    stylesheet: tidy.xsl
- store: 
     directory: /tmp/
- finalize:
     cacheDuration: PT4H
     validUntil: P10D
     ID: SignTestPortalverbundGvAt_
- sign:
    key: pkcs11:///usr/lib64/libeToken.so/test
- publish: /var/lib/git/pvmd/published/metadata_signed.xml
```

**Comment by leifj** [comment]
ok so I think the problem is that since load-from-directory doesn't validate incoming signatures, those are also not stripped. The sign codepath detects the signature element and tries to treat it as a "to-be-signed"  template instead of adding a new signature template. There are multiple issues that needs to get fixed: #88 and #89

**Comment by leifj** [comment]
I think I have a fix but ... that example will actually never work in practice. The XADES signature has Reference@URI="", i.e an implicit same-document reference. The default template in pyXMLSecurity is also an implicit same-document reference and there is no way you can get two different such signatures to verify in an inter-operable way across multiple implementations. The fix is useful anyway.

**Comment by leifj** [comment]
Fixed over in leifj/pyXMLSecurity@ef0534d

**Comment by leifj** [comment]
Your pipeline would need to change to this:

```yaml
- when cleanup:
   - prune:
      - .//{http://www.w3.org/2000/09/xmldsig#}Signature
- when batch:
   - load:
     - /var/lib/pyff/input via cleanup
   - select
   - xslt:
       stylesheet: tidy.xsl
   - store: 
        directory: /tmp/
   - finalize:
        cacheDuration: PT4H
        validUntil: P10D
        ID: SignTestPortalverbundGvAt_
   - sign:
       key: pkcs11:///usr/lib64/libeToken.so/test
   - publish: /var/lib/git/pvmd/published/metadata_signed.xml
```
